src/main/scala/freeeventsourcing/syntax/ProcessSyntax.scala

```diff
@@ -0,0 +1,218 @@
+          case Inr(otherEvent) ⇒ t.effectFor(paths.tail)(otherEvent).map(Inr(_))
+        }
+        def alternatives(paths: P :: T) = {
+          def alt[A, M <: Alternatives[BC]](await: Await[BC, A], more: M) = Alternative[BC, A, M](await, more)
+          alt(paths.head.action, t.alternatives(paths.tail))
+        }
+      }
+    }
+
+    final class FirstOfBuilder[A <: HList] private[ProcessSyntax] (paths: A) {
+      /** Await an event using a selector. The event must be from this bounded context. */
+      def on[S <: WithEventType: EventSelector: ValidSelector](selector: S) = new OnBuilder(selector)
+
+      /** Await event from a specific aggregate. */
+      def from[Id, A](aggregate: Id)(implicit afi: AggregateFromId[Id, BC#Aggregates]) = {
+        val aggregateType = afi.aggregate(boundedContext.aggregates)
+        new FromAggregateBuilder[afi.Out](aggregateType, aggregate)
+      }
+
+      /** Wait until the specified instant, then return unit. */
+      def timeout[R](when: Instant)(thenDo: ProcessMonad[R]): FirstOfBuilder[SwitchPath.Aux[Unit, R] :: A] = {
+        val path = new SwitchPath {
+          type Event = Unit
+          type Result = R
+          val action = WaitUntil[BC](when)
+          def effect(event: Unit) = thenDo
+
+        }
+        new FirstOfBuilder(path :: paths)
+      }
+
+      private[ProcessSyntax] def collect: A = paths
+
+      final class OnBuilder[S <: WithEventType: EventSelector: ValidSelector] private[FirstOfBuilder] (selector: S) {
+        /** This is a flatMap */
+        def execute[R](body: S#Event ⇒ ProcessMonad[R]) = flatMap(body)
+
+        def map[R](f: S#Event ⇒ R) = flatMap(f.andThen(Monad[ProcessMonad].pure))
+
+        def event = map(identity)
+
+        /** Terminate the process if the event occurs. */
+        def terminate = flatMap(_ ⇒ ProcessSyntax.this.terminate)
+
+        def flatMap[R](body: S#Event ⇒ ProcessMonad[R]): FirstOfBuilder[SwitchPath.Aux[S#Event, R] :: A] = {
+          val path = new SwitchPath {
+            type Event = S#Event
+            type Result = R
+            val action = AwaitEvent[BC, S](OnBuilder.this.selector)
+            def effect(event: Event) = body(event)
+          }
+          new FirstOfBuilder(path :: paths)
+        }
+      }
+
+      final class FromAggregateBuilder[A <: Aggregate] private[FirstOfBuilder] (aggregateType: A, aggregate: A#Id) {
+        /** Await an event from the aggregate. */
+        def on[E <: A#Event: AggregateEventType](implicit ev: ValidAggregate[BC, A], idser: StringSerializable[A#Id]) = {
+          val selector = AggregateEventSelector(aggregateType)(aggregate)[E]
+          FirstOfBuilder.this.on(selector)
+        }
+      }
+    }
+
+    /** The same as C =:= CNil, but with a nicer error message. */
+    @implicitNotFound("Not all possible errors of the command were handled, ${C} are still unhandled")
+    sealed trait AllErrorsHandled[C <: Coproduct]
+    object AllErrorsHandled {
```

Comment:
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [Object name does not match the regular expression '"[A-Z][A-Za-z]*"'](https://www.codacy.com/app/msiegenthaler/free-event-sourcing/file/2571398663/issues/source?bid=3298199&fileBranchId=3298202#l211)